design/backup-resources-order.md

```diff
@@ -0,0 +1,39 @@
+## Backup Resources Order
+This document proposes a solution that allows user to specify a backup order for resources of specific resource type.
+
+## Background
+During backup process, user may need to back up resources of specific type in some specific order to ensure the resources were backup properly because these resources are related and ordering might be required to preserve the consistency for the apps to recover itself �from the backup image 
+(Ex: primary-secondary database pods in a cluster).
+
+## Goals
+- Enable user to specify an order of back up resources belong to specific resource type
+
+## Alternatives Considered
+- Use a plugin to backup an resources and all the sub resources.  For example use a plugin for StatefulSet and backup pods belong to the StatefulSet in specific order.  This plugin solution is not generic and requires plugin for each resource type.
+
+## High-Level Design
+User will specify a map of resource type to list resource names (separate by semicolons).  Each name will be in the format "namespaceName/resourceName" to enable ordering accross namespaces.  Based on this map, the resources of each resource type will be sorted by the order specified in the list of resources.  If a resource instance belong to that specific type but its name is not in the order list, then it will be put behind other resources that are in the list.

```

Comment:
I think this proposal works well for pods that are managed by a statefulset and may not be very well suited for pods managed by deployments or daemonsets... As the pods will have generated names. I think it may be worth calling that out in the `Goals` section of the proposal.
I mention pods here, but any resources with a generated name will have the same issue.

Reply:
All the pods' name will be explicitly specified in the "OrderedResources" so it should not matter whether the pods belong to StatefulSet or Deployment or, DaemonSet etc. 
In my usage, the controller code will fetch all pods with specific label (say "app:mysql") and preprocessed them to generate a sequence of podNames to pass to Velero Backup's OrderredResources".

Reply:
The point I was trying to make was that pods that are controlled by a statefulset have predicable names, e.g.`prometheus-1`, `prometheus-2`. However, for pods controlled by deployments and daemonsets, their names will be generated, e.g. `restic-jnxj4` and `velero-78d94698d6-rh8hc`.

Furthermore, if pod `pometheus-1`, controlled by a statefulset, were to crash, then the pod replacing it will also be named `prometheus-1` however the same does not happen with pods xontrolled by deployments or daemonsets. if pod `velero-78d94698d6-rh8hc` were to crash, then its replacement will be a pod whose name will be `velero-<random>`.

Considering that pods, which are ephemeral resources, can be replaced and that their names can change, listing pods controlled by deployments or daemonsets may not work as seamlessly as it would for statefulsets.
I was hoping that this could be called out explicitly in the goals of the proposal.

Reply:
Thanks.  I called out this limitation in the backup-resources-order.md.